### ISSUE TITLE
Problem playing last item: [Errno 32] Broken pipe

### ISSUE THREAD

**Issue by jose1711** [mps-youtube/mps-youtube#706]
hello,

i am getting the error in %subj while trying to play any video. the playback starts in the background but i am unable to control the player.

OS: slackware-current. both mps-youtube and youtube_dl installed using pip3. any idea how to resolve/debug it?

edit: actually i *am* able to control the player, only the control hints are not displayed

thank you, j

**Comment by ritiek** [comment]
What player are you using?

**Comment by jose1711** [comment]
mpv

**Comment by y1n0** [comment]
I have the same problem. I'm on Arch and it happens only after the last youtube-dl upgrade.

**Comment by jose1711** [comment]
i don't know. could be something else. also, on my arch linux box i could not reproduce this problem (youtube-dl 2017.10.29)

**Comment by ngld** [comment]
Installing `python-gobject` on Arch Linux fixed the problem.

I encountered this problem as well and managed to get a traceback in debug mode:
```
raceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/mps_youtube/main.py", line 64, in matchfunction
    func(*matches)
  File "/usr/lib/python3.6/site-packages/mps_youtube/commands/play.py", line 85, in play
    play_range(songlist, shuffle, repeat, override)
  File "/usr/lib/python3.6/site-packages/mps_youtube/player.py", line 41, in play_range
    returncode = _playsong(song, override=override)
  File "/usr/lib/python3.6/site-packages/mps_youtube/player.py", line 215, in _playsong
    returncode = _launch_player(song, songdata, cmd)
  File "/usr/lib/python3.6/site-packages/mps_youtube/player.py", line 443, in _launch_player
    g.mprisctl.send(('stop', True))
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 404, in _send_bytes
    self._send(header + buf)
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 368, in _send
    n = write(self._handle, buf)
BrokenPipeError: [Errno 32] Broken pipe
```

Which was caused by
```
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.6/site-packages/mps_youtube/mpris.py", line 522, in main
    mprisctl = Mpris2Controller()
  File "/usr/lib/python3.6/site-packages/mps_youtube/mpris.py", line 56, in __init__
    from gi.repository import GLib
ModuleNotFoundError: No module named 'gi.repository'
```

It'd be nice if we could get a proper error message that a dependency is missing.

**Comment by jose1711** [comment]
i get no error on import, yet still have the broken pipe error on slackware

**Comment by sz-lhl** [comment]
Thank you ，i will it




发自网易邮箱大师
On 11/10/2017 21:07, ngld wrote:

Installing python-gobject on Arch Linux fixed the problem.

I encountered this problem as well and managed to get a traceback in debug mode:

raceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/mps_youtube/main.py", line 64, in matchfunction
    func(*matches)
  File "/usr/lib/python3.6/site-packages/mps_youtube/commands/play.py", line 85, in play
    play_range(songlist, shuffle, repeat, override)
  File "/usr/lib/python3.6/site-packages/mps_youtube/player.py", line 41, in play_range
    returncode = _playsong(song, override=override)
  File "/usr/lib/python3.6/site-packages/mps_youtube/player.py", line 215, in _playsong
    returncode = _launch_player(song, songdata, cmd)
  File "/usr/lib/python3.6/site-packages/mps_youtube/player.py", line 443, in _launch_player
    g.mprisctl.send(('stop', True))
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 404, in _send_bytes
    self._send(header + buf)
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 368, in _send
    n = write(self._handle, buf)
BrokenPipeError: [Errno 32] Broken pipe


Which was caused by

Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.6/site-packages/mps_youtube/mpris.py", line 522, in main
    mprisctl = Mpris2Controller()
  File "/usr/lib/python3.6/site-packages/mps_youtube/mpris.py", line 56, in __init__
    from gi.repository import GLib
ModuleNotFoundError: No module named 'gi.repository'


It'd be nice if we could get a proper error message that a dependency is missing.

**Comment by johanbluecreek** [comment]
Hi,

I'm also experiencing the "`Problem playing last item: [Errno 32] Broken pipe`"-error. This is on Gentoo after a switch from Python 3.4 to 3.5 in the portage package manager.

> I have the same problem. I'm on Arch and it happens only after the last `youtube-dl` upgrade.

My `youtube-dl` was upgraded as well, but reverting to the previous version (that is to the version used before the switch to 3.5) still reproduce the `[Errno 32]`-error.

> Installing `python-gobject` on Arch Linux fixed the problem.

For me, `dev-python/pygobject` is installed and was not upgraded. This is however not listed as a dependency in Gentoo (at least not a direct dependency).

I also took a look at `dev-python/pafy` which is a direct dependency and has not been upgraded, so only affected by the 3.5 switch.

Running `mpsyt --debug` produce the same/very similar traceback as previously posted

```
Traceback (most recent call last):
  File "/usr/lib64/python3.5/site-packages/mps_youtube/main.py", line 64, in matchfunction
    func(*matches)
  File "/usr/lib64/python3.5/site-packages/mps_youtube/commands/play.py", line 85, in play
    play_range(songlist, shuffle, repeat, override)
  File "/usr/lib64/python3.5/site-packages/mps_youtube/player.py", line 41, in play_range
    returncode = _playsong(song, override=override)
  File "/usr/lib64/python3.5/site-packages/mps_youtube/player.py", line 215, in _playsong
    returncode = _launch_player(song, songdata, cmd)
  File "/usr/lib64/python3.5/site-packages/mps_youtube/player.py", line 443, in _launch_player
    g.mprisctl.send(('stop', True))
  File "/usr/lib64/python3.5/multiprocessing/connection.py", line 206, in send
    self._send_bytes(ForkingPickler.dumps(obj))
  File "/usr/lib64/python3.5/multiprocessing/connection.py", line 404, in _send_bytes
    self._send(header + buf)
  File "/usr/lib64/python3.5/multiprocessing/connection.py", line 368, in _send
    n = write(self._handle, buf)
BrokenPipeError: [Errno 32] Broken pipe

Problem playing last item: [Errno 32] Broken pipe
```

Everything is from the package manager in Gentoo, no `pip` packages.

**Comment by jose1711** [comment]
let me share my debug log too.
```
# mpsyt --logging --debug
mpsyt version      : 0.2.7.1
   notes           : released 6 July 2016
pafy version       : 0.5.3.1 (youtube-dl backend)
youtube-dl version : 2017.10.15.1
Python version     : 3.6.3 (default, Oct  5 2017, 17:06:58) 
[GCC 7.2.0]
Processor          : 05/08
Machine type       : i586
Architecture       : 32bit, ELF
Platform           : Linux-4.12.1-i586-05-08-with-slackware-14.2
sys.stdout.enc     : UTF-8
default enc        : utf-8
Config dir         : /root/.config/mps-youtube
env:TERM           : linux
env:SHELL          : /bin/bash
env:LANG           : en_US.UTF-8
mpris interface couldn't be initialized. Is dbus properly configured?
--

Enter /search-term to search or [h]elp  
..
..
(search & select an item to play)
 fetching item..                                                                            YqgPyqyh4X4; webm 160k; 3 Mb                                                              --                                        

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/mps_youtube/main.py", line 64, in matchfunction
    func(*matches)
  File "/usr/lib/python3.6/site-packages/mps_youtube/commands/play.py", line 85, in play
    play_range(songlist, shuffle, repeat, override)
  File "/usr/lib/python3.6/site-packages/mps_youtube/player.py", line 41, in play_range
    returncode = _playsong(song, override=override)
  File "/usr/lib/python3.6/site-packages/mps_youtube/player.py", line 215, in _playsong
    returncode = _launch_player(song, songdata, cmd)
  File "/usr/lib/python3.6/site-packages/mps_youtube/player.py", line 443, in _launch_player
    g.mprisctl.send(('stop', True))
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 404, in _send_bytes
    self._send(header + buf)
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 368, in _send
    n = write(self._handle, buf)
BrokenPipeError: [Errno 32] Broken pipe

Problem playing last item: [Errno 32] Broken pipe                                                                       
> [file] Cannot open file '/tmp/mpsyt-inputh8l7rv43': No such file or directory
Failed to open /tmp/mpsyt-inputh8l7rv43.
[input] Can't open input config file /tmp/mpsyt-inputh8l7rv43.
Playing: https://r1---sn-vg5obxn3-cuns.googlevideo.com/videoplayback?id=o-AMZQDP29PGmL7Y4d4v9p2KKfNUZdhr4CE5fxB0KhY2H_&keepalive=yes&clen=3643650&ip=85.237.234.10&ei=0kYYWoS-IYGd1wKUgqagCg&sparams=clen%2Cdur%2Cei%2Cgir%2Cid%2Cinitcwndbps%2Cip%2Cipbits%2Citag%2Ckeepalive%2Clmt%2Cmime%2Cmm%2Cmn%2Cms%2Cmv%2Cpl%2Crequiressl%2Csource%2Cexpire&ms=au&mt=1511540258&mv=m&mm=31&mn=sn-vg5obxn3-cuns&mime=audio%2Fwebm&key=yt6&expire=1511562034&lmt=1496757664619932&ipbits=0&itag=251&initcwndbps=612500&pl=19&dur=310.121&source=youtube&requiressl=yes&gir=yes&signature=8A2EE3B38CFF83A814706DD2DA3B25D2265A1B31.992A185CBEA648772258B614D0DBFBEA7A4E3B83&ratebypass=yes
 (+) Audio --aid=1 --alang=eng (*) (opus 2ch 48000Hz)
AO: [pulse] 48000Hz stereo 2ch float


Exiting... (Quit)
q
```
an early mention of mpris issue got me thinking it could be somehow related. ultimately i discovered this little workaround made the Broken pipe problem go away. ideally someone with a deeper knowledge looks into it and provide an explanation/real fix. also ymmv as it already seems this error can mask more than one problem.
```
--- /usr/lib/python3.6/site-packages/mps_youtube/player.py      2017-10-18 15:53:22.476034641 +0200
+++ /home/jose/.local/lib/python3.6/site-packages/mps_youtube/player.py 2017-11-23 15:02:03.410746364 +0100
@@ -388,6 +388,7 @@

         elif "mpv" in config.PLAYER.get:
             cmd.append('--input-conf=' + input_file)
+            g.mprisctl = None

             if g.mpv_usesock:
                 sockpath = tempfile.mktemp('.sock', 'mpsyt-mpv')
```

**Comment by hrnr** [comment]
Hmm seems like the MPRIS process crashed (or exited), which lets to SIGPIPE being send to main python process. We need to investigate why the MPRIS process crashed. Is anybody able to create a reproducer in a docker image? Then I can take a look on that.

As a workaround removing package python-dbus (or dbus-python) should also work do disable MPRIS.

**Comment by hrnr** [comment]
Sorry, no need for docker reproducer. I think I know what happened. Thanks for the debug log, it was very helpful.

The issue on Arch is different than issue on gentoo. I will write on the later today.

**Comment by hrnr** [comment]
This should be fixed by #728. Could you please test that? Thanks very much for your effort and provided logs. 

Just for the record: Arch issue was missing glib gi repositry. Gentoo issue is probably that you are not running dbus daemon, which leads mpris process to exit.

All should be fixed.

**Comment by johanbluecreek** [comment]
> This should be fixed by #728. Could you please test that?

Development branch tested on Gentoo, and I no longer get the error. Thank you.

**Comment by hrnr** [comment]
Thanks for verification. I'm glad that we have such a good community.
